SwiftCompilerSources/Sources/Optimizer/ModulePasses/MandatoryPerformanceOptimizations.swift

```diff
@@ -123,9 +129,104 @@ private func shouldInline(apply: FullApplySite, callee: Function) -> Bool {
     // Force inlining them in global initializers so that it's possible to statically initialize the global.
     return true
   }
+  if apply.parentFunction.isGlobalInitOnceFunction,
+      let global = apply.parentFunction.getInitializedGlobal(),
+      global.mustBeInitializedStatically,
+      let applyInst = apply as? ApplyInst,
+      let projectionPath = applyInst.isStored(to: global),
+      !alreadyInlinedFunctions[projectionPath, default: Set()].contains(callee) {
+    alreadyInlinedFunctions[projectionPath, default: Set()].insert(callee)
+    return true
+  }
   return false
 }
 
+private extension Value {
+  /// Analyzes the def-use chain of an apply instruction, and looks for a single chain that leads to a store instruction
+  /// that initializes a part of a global variable or the entire variable:
+  ///
+  /// Example:
+  ///   %g = global_addr @global
+  ///   ...
+  ///   %f = function_ref @func
+  ///   %apply = apply %f(...)
+  ///   store %apply to %g   <--- is a store to the global trivially (the apply result is immediately going into a store)
+  ///
+  /// Example:
+  ///   %apply = apply %f(...)
+  ///   %apply2 = apply %f2(%apply)
+  ///   store %apply2 to %g   <--- is a store to the global (the apply result has a single chain into the store)
+  ///
+  /// Example:
+  ///   %a = apply %f(...)
+  ///   %s = struct $MyStruct (%a, %b)
+  ///   store %s to %g   <--- is a partial store to the global (returned SmallProjectionPath is MyStruct.s0)
+  ///
+  /// Example:
+  ///   %a = apply %f(...)
+  ///   %as = struct $AStruct (%other, %a)
+  ///   %bs = struct $BStruct (%as, %bother)
+  ///   store %bs to %g   <--- is a partial store to the global (returned SmallProjectionPath is MyStruct.s0.s1)
+  ///
+  /// Returns nil if we cannot find a singular def-use use chain (e.g. because a value has more than one user)
+  /// leading to a store to the specified global variable.
+  func isStored(to global: GlobalVariable) -> SmallProjectionPath? {
+    var singleUseValue: any Value = self
+    var path = SmallProjectionPath()
+    while true {
+      guard let use = singleUseValue.uses.singleUse else {
+        return nil
+      }
+      
+      switch use.instruction {
+      case is StructInst:
+        path = path.push(.structField, index: use.index)
+        break
+      case is TupleInst:
+        path = path.push(.tupleField, index: use.index)
+        break
+      case is EnumInst:
+        path = path.push(.enumCase, index: use.index)
```

Comment:
`use.index` is wrong here. The `enum` instruction always has a single operand (in case of a payload). You need to push `EnumInst.caseIndex`.